### PR TITLE
Flash predict step

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds internals support for the new DataPipeline API of Flash.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -348,7 +348,12 @@ def connect_precision_plugin(self, plugin: PrecisionPlugin) -> None:
 
     def to_device(self, batch: Any) -> Any:
         """Pushes the batch to the root device"""
-        return self.batch_to_device(batch, self.root_device)
+        # Todo (tchaton) Better fix
+        is_dict = isinstance(batch, dict)
+        if is_dict:
+            batch = [batch]
```
Comment (kaushikb11) [inline pytorch_lightning/accelerators/accelerator.py:354]:
```suggestion
		batch = [batch] if isinstance(batch, dict) else batch
```

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -389,9 +389,10 @@ def request_dataloader(self, dataloader_fx: Callable) -> DataLoader:
         Returns:
             The dataloader
         """
-        dataloader = dataloader_fx()
+        if model.trainer is not None:
+            model.trainer.call_hook(f"on_request_{stage}_dataloader")
+        dataloader: DataLoader = getattr(model, f'{stage}_dataloader')()
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/data_loading.py:394]: Do we require a check here for `getattr`?

File: tests/trainer/test_dataloaders.py [hunk 3]
```diff
@@ -1159,3 +1159,71 @@ def test_replace_sampler_with_multiprocessing_context(tmpdir):
 
     new_data_loader = trainer.replace_sampler(train, SequentialSampler(train.dataset))
     assert (new_data_loader.multiprocessing_context == train.multiprocessing_context)
+
```
Comment (kaushikb11) [inline tests/trainer/test_dataloaders.py:1162]: Could we add a test of Flash's DataPipeline Mock for the PR?

File: pytorch_lightning/core/datamodule.py [hunk 4]
```diff
@@ -38,6 +39,13 @@ def __call__(cls, *args, **kwargs):
         2. Assures prepare_data() runs on rank 0
         3. Lets you check prepare_data and setup to see if they've been called
         """
+        __flash_special_attr__ = getattr(cls, "__flash_special_attr__", None)
```
Comment (ananthsub) [inline pytorch_lightning/core/datamodule.py:42]: this feels too magical as a hack to make flash work. can we make this more explicit in the interface for what these attributes are?

File: pytorch_lightning/core/hooks.py [hunk 5]
```diff
@@ -594,6 +606,18 @@ def predict_dataloader(self) -> Union[DataLoader, List[DataLoader]]:
...
+
+    def on_request_val_dataloader(self) -> None:
+        """Called before requesting the val dataloader."""
+
+    def on_request_test_dataloader(self) -> None:
+        """Called before requesting the test dataloader."""
+
+    def on_request_predict_dataloader(self) -> None:
+        """Called before requesting the predict dataloader."""
+
```
Comment (ananthsub) [inline pytorch_lightning/core/hooks.py:620]: n00b q: how are these hooks intended to be used?

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1054,7 +1054,7 @@ def test_epoch_end(self, outputs):
                     self.log('final_metric', final_value)
         """
 
-    def predict(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None):
+    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None):
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1057]: this break backwards compatibility right?

File: pytorch_lightning/trainer/states.py [hunk 7]
```diff
@@ -58,7 +58,7 @@ class RunningStage(LightningEnum):
     """
     TRAINING = 'train'
     SANITY_CHECKING = 'sanity_check'
-    VALIDATING = 'validate'
+    VALIDATING = 'val'
```
Comment (ananthsub) [inline pytorch_lightning/trainer/states.py:61]: can we use the same key values across RunningStage and TrainerState? why is it `val` here but `validate` above?

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:354]: I need to unwrap the dictionary from the list. The boolean is used to keep track of this. But a better fix should be found for using transformers library with Lightning.

File: pytorch_lightning/trainer/data_loading.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:394]: Yes, as the key depends on the stage.

File: pytorch_lightning/core/datamodule.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/core/datamodule.py:42]: Yes, it is used to transform class attributes into instance level attributes through a deepcopy. I didn't find a better solution yet. Any idea ?

File: pytorch_lightning/core/hooks.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/core/hooks.py:620]: In Flash, those hooks are used to inject the attach the data_pipeline into the model dataloaders before they are being fetched.

File: pytorch_lightning/core/lightning.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1057]: Yes, but it is intended. The predict API hasn't been added to the documentation yet as it is likely to changed. However, after this PR, it should be final and I will make a PR to add it to the documentation.

File: tests/trainer/test_dataloaders.py (hunk 3 above)
Comment (tchaton) [reply tests/trainer/test_dataloaders.py:1162]: Yes. Good idea !

File: pytorch_lightning/trainer/states.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/states.py:61]: Yes, good catch ! @carmocca Should we have `val` or `validate` ? I can add a mapping within Flash to map validate to val, but we are adding more inconsistency for validation.

File: pytorch_lightning/core/hooks.py (hunk 5 above)
Comment (SeanNaren) [reply pytorch_lightning/core/hooks.py:620]: can't this just become `on_train_dataloader` etc?

File: pytorch_lightning/core/datamodule.py (hunk 4 above)
Comment (SeanNaren) [reply pytorch_lightning/core/datamodule.py:42]: Inherit the object in Flash and add this logic there? This should be kept in flash imo

Comment (tchaton) [reply pytorch_lightning/core/datamodule.py:42]: Definitely worth to explore, but this is part of the DataModule meta class.Not sure how subclassing metaclass works.

File: pytorch_lightning/trainer/states.py (hunk 7 above)
Comment (carmocca) [reply pytorch_lightning/trainer/states.py:61]: It doesn't matter much to use as we use the Enum and not the str

But i'd keep `validate` in `TrainerState` as that is the function name

Now, whether to use `val` or `validate` in `RunningStage`, some people are pushing to standardize it to `val` (https://github.com/PyTorchLightning/pytorch-lightning/issues/6094#issuecomment-782770802). I don't mind either way personally.

File: pytorch_lightning/core/hooks.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/core/hooks.py:620]: Yes, done !

File: pytorch_lightning/trainer/states.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/states.py:61]: Reverted back to validate. It should be changed in another PR if needed.

File: pytorch_lightning/core/datamodule.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/core/datamodule.py:42]: Moved back to Flash.

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py [hunk 8]
```diff
@@ -348,7 +348,12 @@ def connect_precision_plugin(self, plugin: PrecisionPlugin) -> None:
 
     def to_device(self, batch: Any) -> Any:
         """Pushes the batch to the root device"""
-        return self.batch_to_device(batch, self.root_device)
+        # Todo (tchaton) Better fix
+        is_dict = isinstance(batch, dict)
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/accelerator.py:352]: Could you add a comment for this? What issues were you seeing before?

Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:352]: In Flash, for HF transformers batch (dictionaries with "inputs_ids", "labels", ..), I was receiving only the keys after to_device.

**Review by justusschock — APPROVED** [review]

**Review by ananyahjha93 — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py (hunk 8 above)
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py:352]: could you add a test case for this to ensure the move works as expected?

Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:352]: would it be better to have as not a list?

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 9]
```diff
@@ -295,8 +295,8 @@ def validation_step(self, *args, **kwargs):
     def test_step(self, *args, **kwargs):
         return self.lightning_module.test_step(*args, **kwargs)
 
-    def predict(self, *args, **kwargs):
-        return self.lightning_module.predict(*args, **kwargs)
+    def predict_step(self, *args, **kwargs):
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:298]: I see the point with consistency to `train_step`, but would it be closer to what it does as `predict_batch`

File: pytorch_lightning/trainer/data_loading.py [hunk 10]
```diff
@@ -271,7 +271,7 @@ def _reset_eval_dataloader(
         """
         # always get the loaders first so we can count how many there are
         loader_name = f'{mode}_dataloader'
-        dataloaders = self.request_dataloader(getattr(model, loader_name))
+        dataloaders = self.request_dataloader(model, mode)
```
Comment (Borda) [inline pytorch_lightning/trainer/data_loading.py:274]: not maybe in this PR, but maybe have the mode ad our LightEnum

File: pytorch_lightning/trainer/data_loading.py [hunk 11]
```diff
@@ -380,7 +380,7 @@ def reset_predict_dataloader(self, model) -> None:
         if has_loader:
             self.num_predict_batches, self.predict_dataloaders = self._reset_eval_dataloader(model, 'predict')
 
-    def request_dataloader(self, dataloader_fx: Callable) -> DataLoader:
+    def request_dataloader(self, model: LightningModule, stage) -> DataLoader:
```
Comment (Borda) [inline pytorch_lightning/trainer/data_loading.py:383]: add stage type
```suggestion
    def request_dataloader(self, model: LightningModule, stage: str) -> DataLoader:
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/data_loading.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:274]: Yes, you are right. I will do this in another PR.

File: pytorch_lightning/accelerators/accelerator.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/accelerator.py:352]: I will look into how to properly solve this bug in another PR. Might need a bit more digging.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:298]: Yes, it is for consistency for other functions which also operate on batch.

File: pytorch_lightning/trainer/data_loading.py [hunk 12]
```diff
@@ -389,9 +389,10 @@ def request_dataloader(self, dataloader_fx: Callable) -> DataLoader:
         Returns:
             The dataloader
         """
-        dataloader = dataloader_fx()
+        if model.trainer is not None:
+            model.trainer.call_hook(f"on_{stage}_dataloader")
+        dataloader: DataLoader = getattr(model, f'{stage}_dataloader')()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/data_loading.py:394]: @tchaton why does this call the model's trainer instead of self.call_hook?
